static/app/utils/useLocalStorageState.ts

```diff
@@ -0,0 +1,157 @@
+    // @ts-expect-error
+    return defaultValueOrInitializeFn(value, rawValue);
+  }
+  return defaultValueOrInitializeFn;
+}
+
+// Initialize state with default value or value from localStorage.
+// If window is not defined uses the default value and **does not** throw an error
+function initializeStorage<S>(
+  key: string,
+  defaultValueOrInitializeFn: S | ((rawStorageValue?: unknown) => S)
+): S {
+  if (typeof key !== 'string') {
+    throw new TypeError('useLocalStorage: key must be a string');
+  }
+
+  // Return default if env does not support localStorage. Passing null to initializer
+  // to mimick not having any previously stored value there.
+  if (!SUPPORTS_LOCAL_STORAGE) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, null);
+  }
+
+  // getItem and try and decode it, if null is returned use default initializer
+  const jsonEncodedValue = localStorage.getItem(key);
+  if (jsonEncodedValue === null) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, null);
+  }
+
+  // We may have failed to parse the value, so just pass it down raw to the initializer
+  const decodedValue = tryParseStorage<S>(jsonEncodedValue);
+  if (decodedValue === null) {
+    return defaultOrInitializer(defaultValueOrInitializeFn, undefined, jsonEncodedValue);
+  }
+
+  // We managed to decode the value, so use it
+  return defaultOrInitializer(defaultValueOrInitializeFn, decodedValue, jsonEncodedValue);
+}
+
+// Mimicks the behavior of React.useState but keeps state synced with localStorage.
+// The only difference from React is that when a state initializer fn is passed,
+// the first argument to that function will be the value that we decoded from localStorage
+// and the second argument will be the raw value from localStorage. This is useful for cases where you may
+// want to recover the error, apply a transformation or use an alternative parsing function.
+export function useLocalStorageState<S>(
+  key: string,
+  initialState: S | ((value?: unknown, rawValue?: unknown) => S)
+): [S, (value: S) => void] {
+  const [value, setValue] = useState(() => {
+    return initializeStorage<S>(key, initialState);
+  });
+
+  // We want to avoid a blinking state with the old value when props change, so we reinitialize the state
+  // before the screen updates using useLayoutEffect vs useEffect.
+  useLayoutEffect(() => {
+    setValue(initializeStorage(key, initialState));
+
+    // We only want to update the value when the key changes
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [key]);
+
+  const setStoredValue = useCallback(
+    (newValue: S) => {
+      if (typeof key !== 'string') {
+        throw new TypeError('useLocalStorage: key must be a string');
+      }
+
+      setValue(newValue);
+
+      // Not critical and we dont want to block anything after this
+      scheduleMicroTask(() => {
+        window.localStorage.setItem(key, stringifyForStorage(newValue));
```

Comment:
@JonasBa should we use our localstorage wrapper? https://github.com/getsentry/sentry/blob/master/static/app/utils/localStorage.tsx

Reply:
Oh great q, I'm actually not sure what the purpose of that wrapper is? Is it so that we do not expose destructive methods like `clear`?

Reply:
@JonasBa Not sure…but I feel like we should be consistent in using it. Functionally be the same minus some error handling I guess.

Reply:
Yeah, I don't mind using it at all btw, I was just wondering why it exists :D

Reply:
Actually, one thing that I would like for that wrapper to expose is the clear method... The reason is that in the tests I call clear on beforeEach to make sure it's not polluting values and it would be confusing if I were to call clear on window.localStorage whereas the hook uses some imported storage. Do you mind if I add clear there?

Reply:
@JonasBa works for me, probably just an oversight because we haven’t used clear yet

Reply:
I've made the change to use the wrapper. I think it's actually checking if storage is safe to use (ie not full). I've made it feature complete and removed the `as Storage` type cast so we have some ++ type safety